### ISSUE TITLE
leech model train is single-GPU: the trainer saturates one A30 at 95% and the node has four

### ISSUE THREAD

**Issue by jayhesselberth** [rnabioco/leech#241]
`ModelTrainer` is single-device by construction: `__init__(..., device: str = "cuda", ...)` then `self.model.to(device)` (`leech/training.py`). There is no `DistributedDataParallel`, no `DataParallel`, no `world_size`, and no torchrun entry point anywhere in the trainer — the multi-GPU keyword hits in the package are all in `signal_refine.py`, `_rust_accel.py`, `features.py` and `preparation/parallel.py`, i.e. the CPU/Rust side.

## The measurement that motivates it

Measured on a live `leech model train` of `TCNDwellResidualLN` over the aa-tRNA charging corpus (leech 0.10.0, one A30 on a 4-A30 node):

```
GPU util 95%    mem-bw util 84%    8.9 / 24.6 GB
20.4 min/epoch  x 15 epochs  =  ~5.1 h per seed
train split 32.9 GB, ~6.72M chunks, batch 1024
```

**The trainer saturates the GPU** — this is compute-bound, not input-starved, so it is the case where scaling out actually pays.

Note the asymmetry, because it is easy to generalise the wrong way: `leech model eval` on the same corpus was separately measured **starving at 53-79% GPU with every DataLoader worker pegged**. Eval is input-bound and would get nothing from more GPUs. This issue is about **train only**.

The cost is paid per seed, and a comparison needs several: three paired seeds is the standard in the consuming repo, so one design question is ~15 GPU-hours. The node has 4 A30s and the partition 12; a training run uses 1.

## Three things that will silently go wrong

These are the reason this is not a two-line `nn.parallel.DistributedDataParallel(model)` change.

**1. The sampler conflicts.** `training.py` builds a `WeightedRandomSampler` for class balancing (the `--oversample-minority` path, ~lines 1675-1700) and sets `shuffle=(train_sampler is None)`. DDP requires each rank to draw a **disjoint shard**, which `DistributedSampler` provides and `WeightedRandomSampler` does not. Wrapping the existing sampler in DDP gives every rank the *same* oversampled draw: the run converges, reports plausible metrics, and has trained on `world_size` copies of one shard. Nothing errors.

The corpus this matters on is ~13% positive, so the weighted path is the one in use, not a corner case.

**2. The checkpoint changes shape.** DDP prefixes every `state_dict` key with `module.`. `model_best.pt` is consumed by `leech model export --format onnx` and, downstream of that, by rnabioco/escapepod-models' registry chain, which ships the ONNX as a model bundle. A prefixed checkpoint either fails to load or — under any `strict=False` path — loads *nothing* and exports an untrained graph. Only rank 0 should write, and it must write the unwrapped module's `state_dict`.

**3. Gradient accumulation already exists and interacts.** `grad_accum_split` splits each batch into N sub-batches, scales each sub-loss by `1/N`, and steps once. Under DDP every `backward()` triggers an allreduce, so accumulation without `model.no_sync()` on the non-final micro-steps pays N× the communication for identical arithmetic. Separately, the effective batch becomes `batch_size x world_size`, and the cosine/warmup schedule is indexed in **steps** — which halve at 2 GPUs. A run that changes `T_max` without saying so is a different recipe, and reads downstream as a model or data effect.

## Acceptance criteria

1. **Numerical equivalence, against a measured bar.** A 1-GPU and a 2-GPU run at the same *effective* batch and the same seed land within the corpus's run-to-run noise. That noise is measured, so this is sharp rather than eyeballed: within-arm AUROC sd is **0.000023** across three paired seeds (n=7,599,113) on this corpus. Report the delta against that number.

2. **Sharding is verified, not assumed.** A test asserting that the union of the ranks' sampled indices covers the epoch exactly once in the unshuffled case, and reproduces the single-GPU class ratio in the weighted case. This is the failure that produces a plausible-looking wrong model, so it needs a test that fails when sharding is absent — not one that passes because both ranks happened to converge.

3. **The exported checkpoint is byte-compatible with the single-GPU one.** Same keys, no `module.` prefix, and `leech model export --format onnx` succeeds from a 2-GPU checkpoint. A key-set equality test between a 1-GPU and a 2-GPU `model_best.pt` is the cheap guard.

4. **Accumulation uses `no_sync()`** on the non-final micro-steps, and the scheduler's step count is corrected for `world_size` (or explicitly documented as not corrected).

5. **Single-GPU behaviour is unchanged.** With `world_size == 1` the code path, the RNG draw and the resulting checkpoint are identical to today's — verified by a same-seed checkpoint comparison, not by inspection.

6. Scaling reported honestly: epoch wall time at 1, 2 and 4 GPUs on the corpus above. If it is 1.6x rather than 2x, say 1.6x.

## Non-goals

- **Not multi-node.** Single node, up to its 4 GPUs.
- **Not FSDP or model parallelism.** The model fits with room to spare (8.9 of 24.6 GB); this is data parallelism only.
- **Not eval.** Measured input-bound (53-79% GPU, workers pegged); DDP buys nothing and the change would add risk to the path that produces the scores.
- **Not a new default.** Single-GPU stays the default; multi-GPU is opt-in via an explicit flag.
- **Not a recipe change.** Batch size, LR and epoch budget stay where they are — the point is the same training, faster, so that existing comparisons remain paired.

**Comment by jayhesselberth** [comment]
## Implementation plan

### 0. The memory question first, because it gates everything

If a second rank does not fit beside the first, none of the rest matters. Measured on
the live run this issue describes — job `277764`, `charging_leech_train` /
`production` / `TCNDwellResidualLN`, `--batch-size 1024 --oversample-minority
--mixed-precision --num-workers 2 --epochs 15 --seed 0`, 12 epochs done in 4 h 18 m
(≈21.5 min/epoch wall, validation included) on one A30 of `compgpu01`:

| | |
|---|---|
| job cgroup `memory.current` | 42,615,107,584 B = **39.7 GiB** |
| job cgroup `memory.peak` | 42,980,777,984 B = **40.0 GiB** |
| of which `anon` | 40,707,874,816 B = **37.9 GiB** |
| `file` / `shmem` | 1.5 GiB / 0.74 GiB |
| allocation (`--mem 64000`) | 62.5 GiB |
| corpus on disk | `train.npz` 31 GiB + `val.npz` 5.5 GiB = 36.5 GiB |
| node `compgpu01` | 250 GiB total, 45 GiB used, 205 GiB available, 4× A30, 64 cores |

`sstat` reports MaxRSS 39.7 GiB and the two `pt_data_worker` processes each show
~35.8 GiB RSS, which looks like 110 GiB until you notice the workers are forks and
their pages are COW-shared with the parent. The cgroup counts a shared page once, so
**39.7 GiB is the honest footprint of the whole process tree**, and 37.9 GiB of it is
anonymous — the corpus, resident, essentially 1:1 with the 36.5 GiB of npz.

That 1:1 is the projection rule, because a spawned DDP rank shares nothing with its
siblings and re-reads the corpus itself:

| ranks | anon | request | fraction of the node |
|---|---|---|---|
| 1 (today) | 37.9 GiB | 62.5 GiB | 16% |
| 2 | ~80 GiB | `mem_mb 88000` | 34% |
| 4 | ~160 GiB | `mem_mb 176000` | 68% |

**It fits.** Two conditions, both worth stating plainly:

- The Snakemake `charging_leech_train:mem_mb=64000` must scale with the GPU count.
  A 2-GPU run inside today's 64 GB allocation is OOM-killed at rank 1's load, not
  slow — so the resource and the flag have to move together.
- At 4 ranks we hold 68% of the node's RAM, and page cache (182 GiB warm right now)
  is squeezed accordingly, so the four concurrent corpus reads land on disk. That is
  a one-time startup cost, not a per-epoch one. It displaces no other GPU job,
  because a 4-GPU run already takes every A30 on the node.

The shared-memory contingency below (one load, `share_memory_()`, ranks map it) would
hold the total at ~40 GiB for any rank count. It is an optimization here rather than
a prerequisite — worth having if we ever want two 2-GPU runs on one node, not worth
blocking on now.

### Two corrections to the issue, both in our favour

1. **The LR schedule is epoch-indexed, not step-indexed.**
   `linear_warmup_cosine_decay(total_epochs=epochs, …)` drives a `LambdaLR` that
   `train()` steps once per epoch (`training.py:1152`), as do the plateau path and
   the manual warmup ramp. Nothing in the trainer counts optimizer steps. Acceptance
   criterion 4's second half therefore has nothing to correct — it needs stating in
   the PR, not fixing.

2. **There is a fourth silent failure the issue does not list.** `AdversarialHead`
   and `RegressionHead` are in the optimizer but live *outside* the model
   (`training.py:462-500`). Under a naive DDP their gradients are never allreduced,
   so every rank keeps a private aux head while the backbone stays in sync. Converges,
   reports plausible numbers, wrong — the same shape as the sampler failure.

### Design decisions

**`--batch-size` stays the global batch, split across ranks** (per-rank =
`batch_size // world_size`). Not the PyTorch convention, and right here: the issue's
own non-goal is "not a recipe change", and this way steps per epoch, the epoch-indexed
LR schedule, `ClipGrad`'s quantile buffer and the `grad_accum_split` arithmetic are
what they are today by construction, while "same effective batch" in criterion 1
becomes *the same command line*. DDP averages per-rank mean losses over equal-size
shards, which is exactly the single-GPU mean.

**Launch: `--gpus N`, self-spawning.** `torch.multiprocessing.spawn` from inside
`train_model`, also honoring `RANK`/`WORLD_SIZE`/`LOCAL_RANK` under `torchrun`. Keeps
`leech model train` the single entry point the pipeline already calls. Backend follows
the device: `nccl` on cuda, `gloo` on cpu — the CPU path is what lets the sharding and
checkpoint tests run in CI without a GPU.

**Checkpoint shape.** `model = torch.compile(model)` already happens at
`training.py:1818`, so checkpoints carry `_orig_mod.` keys today and every consumer
strips exactly that prefix (`model_loading.py:95`, `bundling.py:44`,
`inference/bundle.py:284`). Nothing strips `module.`. Wrapping as
`DDP(torch.compile(model))` and taking `state_dict()` off the unwrapped module keeps
the key set byte-identical.

### Work

**New `src/leech/distributed.py`**
- `DistContext(rank, local_rank, world_size, backend, is_main)`, env detection,
  `spawn_workers`, `barrier`, `destroy`.
- `DistributedWeightedSampler`: every rank draws the **same global multinomial** from
  a CPU generator seeded `seed + epoch`, then takes `idx[rank::world_size]`. Shards
  are disjoint and their union is *exactly* the single-GPU draw, which is what makes
  criterion 2's weighted half provable rather than statistical. (The corpus is
  6.72M chunks, well under `torch.multinomial`'s 2^24 category ceiling — the same
  ceiling `WeightedRandomSampler` already lives under.)
- `all_reduce_tally`, `gather_eval_arrays` (gather `(index, pred, label)`, dedupe the
  sampler's padding).

**`src/leech/training.py`**
- `Trainer` gains `dist: DistContext | None`; wraps the compiled module as
  `DDP(compiled, device_ids=[local_rank])`. `self.model` keeps pointing at the
  unwrapped module for `state_dict`, `deepcopy(_best_model_state)` and
  `enable_repr_capture`'s `getattr(model, "classifier")` — DDP does not forward
  attribute lookup, compile does.
- Each aux head gets its own DDP wrapper (correction 2 above).
- `no_sync()` on non-final micro-steps via an `ExitStack` over model + heads
  (criterion 4).
- `train_epoch`: all-reduce the `_DeviceTally` sums and `seen`.
- `validate`: sharded with `DistributedSampler(shuffle=False)`, gathered, metrics
  computed from the global arrays. Not optional — validation is ~5 min against 20.4
  min of training, so leaving it on rank 0 caps the whole thing near 2× no matter how
  many GPUs are added.
- `train()`: `set_epoch(epoch)` per epoch; rank 0 alone owns `Progress`, console
  output, checkpoints, `config.json`, `metrics.json`, the seed file; barrier before
  exit.
- `train_model`: spawn when `gpus > 1`; rank 0 writes `metrics.json` and the parent
  re-reads it to preserve the return contract.
- Guards that raise: `gpus > 1` with in-memory `train_chunks` (the grid-search path),
  inside a daemonic pool worker, or above `torch.cuda.device_count()`.

**`src/leech/dataset.py`** — `resolve_dataloader_workers(..., local_world_size=1)` so
the auto cap divides the job's CPU allocation instead of each rank claiming all of it.

**`src/leech/models/inference_wrapper.py`** — optional `forward_module` used by
`forward_batch`; `self.model` stays the unwrapped module.

**CLI** — `--gpus` (default 1) through `cli.py` → `commands/train.py` → `train_model`.

**`pipeline/`** — `train_gpus` config knob, and `mem_mb` scaled with it per the table
above.

Everything is behind `world_size > 1`. At `world_size == 1` no process group is created
and no object changes identity — that is criterion 5.

### Known deviation to fix rather than document

`CrossEntropyLoss(weight=…)` with `reduction="mean"` divides by *the sum of the target
weights in the batch*, not the count. Unlike BCE and focal, that is not decomposable:
averaging per-rank means reweights by each shard's class composition. Only multiclass
runs with class weights are affected. `reduction="sum"` normalized by an all-reduced
weight total when `world_size > 1` is the fix, and it is small enough to just do.

### Tests — `tests/test_distributed_training.py`, 2-rank gloo on CPU

| Criterion | Test |
|---|---|
| 2 | `test_sharding_covers_epoch_once` — union of rank index sets == the epoch exactly once, intersection empty. Fails on a non-sharded sampler because it sees duplicates, not because convergence differs. |
| 2 | `test_weighted_shards_match_single_gpu_draw` — `concat(rank draws)` == the single-GPU `WeightedRandomSampler` draw at the same seed; class ratio equal. |
| 1 | `test_one_step_gradients_match` — one optimizer step from identical init, 1 rank × batch 8 vs 2 ranks × batch 4 over the same 8 samples, gradients equal to fp32 tolerance. Sharper and ~10⁴× cheaper than a convergence run. |
| 3 | `test_checkpoint_keys_match_single_gpu` — key-set equality between a 1-rank and a 2-rank `model_best.pt`; no key contains `module.`; `leech model export --format onnx` succeeds from the 2-rank checkpoint. |
| 4 | `test_no_sync_on_non_final_microsteps` — entered on micro-steps `0..N-2`, not on `N-1`. |
| 5 | `test_single_gpu_creates_no_process_group` — `torch.distributed.is_initialized()` False at `--gpus 1`, and two same-seed runs bit-identical. Plus a stash/compare against `main` during development, reported in the PR. |
| — | `test_aux_heads_stay_in_sync` — adversarial and CL-regression params identical across ranks after a step. |

### Measurement (criteria 1 and 6)

Three `sbatch` jobs on `gpu`, memory sized from the table above: epoch wall time at 1,
2 and 4 GPUs, and a full same-seed 1- vs 2-GPU run reporting ΔAUROC against the
0.000023 within-arm sd. ~8 GPU-hours. Baseline to beat: **21.5 min/epoch**, 15 epochs.

### Sequence

`distributed.py` + samplers + their tests (no trainer changes) → trainer/DDP/checkpoint
+ tests → distributed validation and metric reduction → CLI, pipeline resources,
CLAUDE.md section → the runs.

### Non-goals inherited unchanged

Not multi-node, not FSDP, not `eval`, not a new default, not a recipe change.

**Comment by jayhesselberth** [comment]
## Criterion 6: scaling, measured

Production charging corpus (`TCNDwellResidualLN`, 6.72M chunks, 36.5 GiB npz),
2 epochs, one 4×A30 node, the production recipe verbatim except `--epochs 2`.
Same `--seed 0` and the same global `--batch-size 1024` at every GPU count.

| `--gpus` | job wall | speedup | peak RSS | epoch-2 val AUROC |
|---|---|---|---|---|
| 1 | 24:31 | 1.00x | 40.6 GiB | 0.97350 |
| 2 | 16:40 | 1.47x | 88.3 GiB | 0.97310 |
| 4 | 10:53 | **2.25x** | 157.6 GiB | 0.97275 |

**2.25x at 4 GPUs, not 4x.** The wall clock says why: ~3 min of per-rank corpus
load and the final eval do not shard, so they set a floor that the training
phase cannot amortise at 2 epochs. Training time alone scales better —
18:30 → 10:14 → 8:55 — and a real 30-epoch run spends proportionally less in
that fixed cost, so the end-to-end figure should improve with epoch count. I am
reporting the end-to-end number because that is what a queued job costs.

AUROC spans 7.5e-4 across the three. That is **not** a numerical-equivalence
result and should not be read as one: augmentation noise is drawn in the
DataLoader workers, so a different rank count means a different realized
trajectory by construction. The equivalence claim is
`test_two_rank_gradients_match_the_single_rank_gradients` — one step from an
identical init, equal to 1e-6.

## Two defects the hardware found that CPU tests could not

**The ranks were silent.** `setup_logging` runs in the click entry point, which
a spawned rank never reaches, so the `leech` logger had no handler and every
INFO line was dropped — including rank 0's, where the effective batch, the
sampler statistics and the encoding-fallback warning are reported. The run
worked and said nothing about itself.

**DataLoader workers were spawning, not forking.** `multiprocessing.spawn.prepare`
forces a spawned child's default start method to match how it was created
(verified directly: parent `fork`, rank `spawn`), so every DataLoader in every
rank pickled the dataset's stacked tensors through `/dev/shm` instead of
COW-sharing them — voiding the exact invariant those contiguous buffers exist
for.

It surfaced nowhere near the cause, and did not look like memory: shm pages are
charged to the cgroup but not to RSS, so the run died at 176 GiB RSS against a
244 GiB allocation, in the *fourth* rank's validation loader, as
`No space left on device`. Fixing it took the same run to 157.6 GiB and it
completed. Both fixes carry regression tests.

## Memory

~1:1 with the corpus per rank, so `--gpus` and the job's `mem_mb` have to move
together or the second rank is OOM-killed during its load.
`pipeline/workflow/rules/train.smk` scales both off one `train_gpus` key.
On this cluster 4 ranks want close to the whole node: `--mem 250000M`.
